### DISCUSSION TITLE
docs/reference/zmodel-language

### DISCUSSION THREAD

**Discussion by giscus** [zenstackhq/zenstack-docs#288] [Comments]
# docs/reference/zmodel-language

ZModel language references

https://zenstack.dev/docs/reference/zmodel-language

**Comment by justin-hackin** [comment]
I would like to understand better how to create custom functions. I went to the [linked github page](https://github.com/zenstackhq/zenstack/blob/main/packages/schema/src/res/stdlib.zmodel) regarding this. However, none of the functions seem to have code in them. It looks like TS but I'm not sure 🤷🏻‍♂️ Also, I can't find any references to these `@@@` directives following the function in these docs. I suspect custom functions could help me construct complex conditional arguments more succinctly.

**Reply by ymc9** [reply]
Hi, one thing I want to clarify first is the ZModel functions are not directly implemented in TS. They are compiled in different ways by the CLI based. For example, the [startsWith(title, 'foo')](https://zenstack.dev/docs/reference/zmodel-language#startswith) function call is compiled into a partial Prisma query like:

```js
{
  title: {
    startsWith: 'foo'
  } 
}
```

That part of compilation/transformation is open for extension yet. But I'd like to understand the specific goal you want to achieve here. Thanks!

**Comment by justin-hackin** [comment]
> Hi, one thing I want to clarify first is the ZModel functions are not directly implemented in TS. They are compiled in different ways by the CLI based. For example, the [startsWith(title, 'foo')](https://zenstack.dev/docs/reference/zmodel-language#startswith) function call is compiled into a partial Prisma query like:
> 
> ```js
> {
>   title: {
>     startsWith: 'foo'
>   } 
> }
> ```
> 
> That part of compilation/transformation is open for extension yet. But I'd like to understand the specific goal you want to achieve here. Thanks!

Thanks for taking interest in my goals. My access control conditions are getting a bit wild and I was wondering if I could write a custom function for this. i.e. I have a nested permission structure so I end up doing something like this 
```
@@allow('all',
        regions?[admins?[user == auth()]] ||
        parentCalendarGroup.regions?[admins?[user == auth()]] ||
        parentCalendarGroup.parentCalendarGroup.regions?[admins?[user == auth()]] ||
        parentCalendarGroup.parentCalendarGroup.parentCalendarGroup.regions?[admins?[user == auth()]]
    )
```

This is only going to get hairier if/when I introduce administrative priveledges to each level of the group hierarchy. I have already done some code generation for copy-pasting into my queries since that seems to be the only way that the types work properly in prisma 😑. Maybe I should just do that or it might not be worth it  🤷🏻‍♂️ ... the IDE support in Webstorm is pretty great! 👍 Partly just curious about all the functionality of the library.

**Reply by ymc9** [reply]
Got it.

I see you've already found the feature request here: https://github.com/zenstackhq/zenstack/issues/276

Do you think the solution being discussed over there (reducing duplication by delegating permission check to a relation field) will partly solve your problem?

**Comment by PodkopovP** [comment]
Any way to filter something out, based on the count of relations?

I.e. @@deny('all', members[count() < 2])
